worldedit-core/src/main/java/com/sk89q/worldedit/session/request/Request.java

```diff
@@ -24,13 +24,15 @@
 import com.sk89q.worldedit.world.World;
 
 import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
 
 /**
- * Describes the current request using a {@link ThreadLocal}.
+ * Describes the current request using a Thread.name() -> request map(avoid server freeze on scale of 50+ players).
  */
 public final class Request {
 
-    private static final ThreadLocal<Request> threadLocal = ThreadLocal.withInitial(Request::new);
+    private static Map<String, Request> reqMap = new HashMap<>();
```

Comment:
HashMap is also not thread-safe

Reply:
So how about changing the key of this HashMap?
like thread id or something

Reply:
Do you know what "thread-safe" means?

Reply:
considering the visibility, I think if we can find a key to identify each thread, it's no need to achieve full thread safe in this case.